cpp/oneapi/dal/table/backend/csr_kernels.hpp

```diff
@@ -77,4 +81,85 @@ void csr_pull_block(const Policy& policy,
+                                    const std::vector<sycl::event>& dependencies);
+#endif
+
+/// The number of non-zero elements in the table calculated from the row offsets array in CSR format.
+///
+/// This function dispatches the execution:
+///     If data parallel policy is enabled and there is a sycl queue associated with row offsets array,
+///         then DPC++ implementation of the function is called.
+///     Otherwise, C++ implementation is called.
+///
+/// @param[in] row_offsets  Row offsets array in CSR layout
+///
+/// @return The number of non-zero elements in CSR table
+std::int64_t csr_get_non_zero_count(const array<std::int64_t>& row_offsets);
+
+#ifdef ONEDAL_DATA_PARALLEL
+
+/// Checks that the elements in the input array are not descending
+///
+/// @tparam T   The type of elements in the input array
+///
+/// @param[in] arr          Input array
+/// @param[in] dependencies Events indicating availability of the `arr` for reading.
+///
+/// @return true, if the elements in the array are not descending;
+///         false, otherwise
+template <typename T>
+bool is_sorted(const array<T>& arr, const std::vector<sycl::event>& dependencies);
+
+#endif
+
+/// Given the array A[0], ..., A[n-1] and two values: `min_value` and `max_value`,
+/// checks that min_value <= A[i] <= max_value for each i = 0, ..., n-1.
+///
+/// @tparam T   The type of elements in the input array
+///
+/// @param[in] arr       Input array
+/// @param[in] min_value The lower boundary for the values in the input array
+/// @param[in] max_value The upper boundary for the values in the input array
+///
+/// @return less_than_min,    if there exists i, 0 <= i <= n-1: A[i] < min_value;
+///         within_bounds,    if min_value <= A[i] <= max_value for each i = 0, ..., n-1;
+///         greater_than_max, if there exists i, 0 <= i <= n-1: A[i] > max_value.
+template <typename T>
+out_of_bound_type check_bounds(const array<T>& arr, const T& min_value, const T& max_value);
+
+#ifdef ONEDAL_DATA_PARALLEL
+
+/// Given the array A[0], ..., A[n-1] and two values: `min_value` and `max_value`,
+/// checks that min_value <= A[i] <= max_value for each i = 0, ..., n-1.
+///
+/// @tparam T   The type of elements in the input array
+///
+/// @param[in] arr          Input array
+/// @param[in] min_value    The lower boundary for the values in the input array
+/// @param[in] max_value    The upper boundary for the values in the input array
+/// @param[in] dependencies Events indicating availability of the `arr` for reading.
+///
+/// @return less_than_min,    if there exists i, 0 <= i <= n-1: A[i] < min_value;
+///         within_bounds,    if min_value <= A[i] <= max_value for each i = 0, ..., n-1;
+///         greater_than_max, if there exists i, 0 <= i <= n-1: A[i] > max_value.
+template <typename T>
+out_of_bound_type check_bounds(const array<T>& arr,
+                               const T& min_value,
+                               const T& max_value,
```

Comment:
You can pass values directly in case of primitive types.

```suggestion
                               T min_value,
                               T max_value,
```

Reply:
Wanted to make template code more generic. But Ok, I'll make the change.